src/eigsolve/lanczos.jl

```diff
@@ -150,3 +150,143 @@ function eigsolve(A, x₀, howmany::Int, which::Selector, alg::Lanczos;
+            by, rev = eigsort(which)
+            p = sortperm(D; by=by, rev=rev)
+            D, U = permuteeig!(D, U, p)
+
+            # Detect convergence by computing the residuals
+            bs_r = fact.r_size   # The block size of the residual (decreases as the iteration goes)
+            r = residual(fact)
+            UU = U[(end - bs_r + 1):end, :]  # The last bs_r rows of U, used to compute the residuals
+            normresiduals = let R = block_inner(r, r)
+                map(u -> sqrt(real(dot(u, R, u))), cols(UU))
+            end
+            converged = 0
+            while converged < K && normresiduals[converged + 1] <= tol
+                converged += 1
+            end
+            if converged >= howmany || β <= tol  # Successfully find enough eigenvalues.
+                break
+            elseif verbosity >= EACHITERATION_LEVEL
+                @info "BlockLanczos eigsolve in iteration $numiter, Krylov dimension = $K: $converged values converged, normres = $(normres2string(normresiduals[1:howmany]))"
+            end
+        end
+
+        if K < krylovdim
+            expand!(iter, fact; verbosity=verbosity)
+            numops += 1
+        else # Shrink and restart following the shrinking method of `Lanczos`.
+            numiter >= maxiter && break
+            keep = max(floor(Int, (3 * krylovdim + 2 * converged) / (5 * bs)), 1) * bs
+            H = zeros(S, keep + bs, keep)
+            # The last bs rows of U contribute to calculate errors of Ritz values.
+            @inbounds for j in 1:keep
+                H[j, j] = D[j]
+                H[(keep + 1):end, j] = U[(K - bs + 1):K, j]
+            end
+            # Turn diagonal matrix D into a block tridiagonal matrix, and make sure 
+            # The residual of krylov subspace keeps the form of [0,..,0,R]
+            @inbounds for j in keep:-1:1
+                h, ν = householder(H, j + bs, 1:j, j)
+                H[j + bs, j] = ν
+                H[j + bs, 1:(j - 1)] .= zero(eltype(H))
+                lmul!(h, H)
+                rmul!(view(H, 1:(j + bs - 1), :), h')
+                rmul!(U, h')
+            end
+            # Transform the basis and update the residual and update the BTD.
+            BTD .= S(0)
+            BTD[1:keep, 1:keep] .= H[1:keep, 1:keep]
+            B = basis(fact)
+            basistransform!(B, view(U, :, 1:keep))
+
+            r_new = OrthonormalBasis(fact.r.vec[1:bs_r])
+            view_U = view(U, (K - bs_r + 1):K, (keep - bs_r + 1):keep)
+            basistransform!(r_new, view_U)
+            fact.r.vec[1:bs_r] = r_new[1:bs_r]
+
+            while length(fact) > keep
+                pop!(fact.V)
+                fact.k -= 1
+            end
+            numiter += 1
+        end
+    end
+
+    howmany_actual = howmany
+    if converged > howmany
+        howmany_actual = converged
+    elseif length(D) < howmany
+        howmany_actual = length(D)
+    end
+    U1 = view(U, :, 1:howmany_actual)
```

Comment:
In analogy with the `Lanczos` implementation I would define values here as

```suggestion
    values = D[1:howmany_actual]
    U1 = view(U, :, 1:howmany_actual)
```

and then drop the slicing in the return statement.